### DISCUSSION TITLE
session.load() error handling

### DISCUSSION THREAD

**Discussion by Casper-Guo** [theOehrly/Fast-F1#536] [Q&A]
I have the following code:

```
for i in missing_rounds:
        race = f.get_session(season, i, "R")

        try:
            race.load(telemetry=False)
        except:
            # TODO: Proper handling of FastF1 errors
            logging.warning("Cannot load %s", race)
```

Not critical but I would like to catch specific errors instead of using a wildcard except. Looking through the code the following exceptions are possible:

- `_api.SessionNotAvailableError`
- `core.NoLapDataError`
- Two Ergast exceptions inherited from `interface.ErgastError`

The former two inherits from `Exception` directly. Thoughts about a nice and concise solution for dealing with all these exception classes?

**Comment by theOehrly** [comment]
I'm aware that the API around exceptions in FastF1 is far from optimal and certainly not consistent. A perfect example for this is that you missed the ``RateLimitExceededError`` in ``fastf1.req`` (which even has a TODO comment that it should be documented...).

I don't have any concrete ideas for a consistent solution here. But there are some things that would ideal:
- all exceptions can be imported from a single submodule, either directly from ``fastf1`` or from something like ``fastf1.exceptions``
- all exceptions are documented
- all exceptions use reasonable base classes: that might be ``Exception`` for some, a more specific default exception like ``ValueError`` for others or another more general custom exception
- the exceptions are logically structured and extensible (similar to how the two Ergast exceptions inherit from a common base class)

This has been somewhere on the back of my mind, but I never got around to working on it. I'm open to a PR that overhauls the whole exceptions stuff in FastF1 completely. That's probably a bit of work, though. Alternatively, a well-thought-out plan on how to improve the current situation, which (additional) exceptions might be necessary, which base classes to use, ... would already be a good step in the right direction.

**Reply by Casper-Guo** [reply]
For what it is worth, I did come across `RateLimitExceededError` in my search. I only talked about the other ones because I was only interested in additional handling for errors that directly relate to data availability if that makes sense.

I will port this discussion to an issue.

**Comment by theOehrly** [comment]
> The former two inherits from `Exception` directly. Thoughts about a nice and concise solution for dealing with all these exception classes?

Apart from using 

```
....
except (SessionNotAvailableError, NoLapDataError, ErgastError):
...
```
or multiple individual excepts to catch the errors specifically with different error handling, I don't really have a good suggestion.
